broker-core/src/main/java/io/zeebe/broker/workflow/state/VariablesState.java

```diff
@@ -114,12 +119,29 @@ private void setVariableLocal(
       DirectBuffer value,
       int valueOffset,
       int valueLength) {
-    this.scopeKey.wrapLong(scopeKey);
-    variableNameView.wrap(name, nameOffset, nameLength);
-    this.variableName.wrapBuffer(variableNameView);
 
-    variable.wrapValue(value, valueOffset, valueLength);
-    variablesColumnFamily.put(scopeKeyVariableNameKey, variable);
+    newVariable.wrapValue(value, valueOffset, valueLength);
+    final DirectBuffer newValue = newVariable.getValue();
+
+    final DirectBuffer currentValue = getVariableLocal(scopeKey, name, nameOffset, nameLength);
+
+    if (currentValue == null) {
+      variablesColumnFamily.put(scopeKeyVariableNameKey, newVariable);
+
+      if (listener != null) {
+        listener.onCreate(variableName.getBuffer(), newValue, scopeKey);
+      }
+
+    } else if (!BufferUtil.equals(currentValue, newValue)) {
```

Comment:
If we have here a performance problem I will say: I told you. :smile:

Reply:
I agree but I think we should take it as a future improvement.

Reply:
Do we have an issue for this? :sweat_smile: